Base images for the Forrestal, Supercarriers, Vinson, Kuznetsov, Kuznetsov (2017) and Type 071 Yuzhao

MetalStormGhost opened the pull request:
Added separate base images for the Forrestal, Supercarrier (CVN-75), Kuznetsov, Kuznetsov (2017/Supercarrier) and Type 071 Yuzhao.

Resolves #1292

MetalStormGhost committed:
Added separate base images for the Forrestal, Supercarrier (CVN-75), Kuznetsov, Kuznetsov (2017/Supercarrier) and Type 071 Yuzhao.

Resolves #1292

MetalStormGhost committed:
Added separate base images for the Forrestal, Supercarrier (CVN-75), Kuznetsov, Kuznetsov (2017/Supercarrier) and Type 071 Yuzhao.

Resolves #1292

DanAlbert commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -241,9 +241,47 @@ def closeEvent(self, close_event: QCloseEvent):
+                            return "./resources/ui/carrier_cv59.png"
+                        elif (
+                            unit.type == "Stennis"
+                            and game is not None
```
Why would this ever be `None`?

MetalStormGhost replied:
I encountered at least one instance during testing when game was None during testing. I did not investigate it further, just added a check against it.

DanAlbert replied:
That sounds like a bug. All this does is hide that bug.

MetalStormGhost replied:
Do you want me to take the evaluation against None and the AttributeError checking above it away?

DanAlbert replied:
Unless you can explain to me how this window is open without a game loaded, yes. That should be impossible. If it's happening that is a different bug that should be fixed.

DanAlbert commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -241,9 +241,47 @@ def closeEvent(self, close_event: QCloseEvent):
+        try:
+            game = self.game_model.game
+        except:
+            game = None
```
Huh? Why?

MetalStormGhost replied:
Same as above.

DanAlbert commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -241,9 +241,47 @@ def closeEvent(self, close_event: QCloseEvent):
+            for ground_object in self.cp.ground_objects:
+                for group in ground_object.groups:
+                    for unit in group.units:
+                        if unit.type == "Forrestal":
```
This should be factored out in `ControlPoint`. `ControlPoint.get_carrier_group_type`. Also factor out the common portion of that and `ControlPoint.get_carrier_group_name` into `ControlPoint.get_carrier_group`.

MetalStormGhost replied:
Done. In Revision: 4de3fee2a32c2be3819ee8eb894f72d7c4795d37

MetalStormGhost replied:
This also ended up fixing a bug in flightgroupspawner.py where mission.find_group(group_name) would fail, causing the mission generation to also not succeed.
![image](https://user-images.githubusercontent.com/89945461/145716705-50085c3b-5f72-49c3-b9b4-1df78a4ba6e5.png)

DanAlbert commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -241,9 +241,47 @@ def closeEvent(self, close_event: QCloseEvent):
+                            or unit.type == "CVN_71"
+                            or unit.type == "CVN_72"
+                            or unit.type == "CVN_73"
+                            or unit.type == "CVN_75"
```
This is getting messy. This sort of thing should be factored out into unit data yaml files. We don't currently have unit data yaml for naval units, but those should be added. For now just a mapping from unit type to the banner is fine.

Example `resources/units/ships/CVN_75.yaml`:
```yaml
---
control_point_banner: resources/ui/carrier_cvn75.png
```

DanAlbert replied:
Alternatively, just use `resources/ui/units/ships/{unit.type}.png` and add an image per unit type (they can all be copies if you don't want to track down unique images for each ship right now).

MetalStormGhost replied:
I added images for the rest of the carriers and moved them to resources/ui/units/ships/ as requested.

MetalStormGhost committed:
Implemented controlpoint.get_carrier_group() and refactored QBaseMenu2 to use it in order to clean up base image evaluation. Added several supercarrier unit images and moved the existing ones to resources/ui/units/ships/. Also added support for the Vinson carrier type.

MetalStormGhost committed:
Implemented controlpoint.get_carrier_group() and refactored QBaseMenu2 to use it in order to clean up base image evaluation. Added several supercarrier unit images and moved the existing ones to resources/ui/units/ships/. Also added support for the Vinson carrier type.

MetalStormGhost committed:
Some cleaning up in QBaseMenu2.py

MetalStormGhost committed:
Removed AttributeError for null game in get_base_image().

DanAlbert commented on game/db.py:
```diff
@@ -330,7 +331,7 @@
 
 def upgrade_to_supercarrier(unit: Type[ShipType], name: str) -> Type[ShipType]:
-    if unit == Stennis:
+    if unit == Stennis or unit == VINSON:
```
This is impossible and just future-proofing, right? I don't understand how this relates to the PR.

MetalStormGhost replied:
Indeed, future-proofing.

DanAlbert replied:
Drop all this from this PR and add it to that one IMO. We don't know that this is the structure the other PR will end up with, and right now it's just dead code that confuses the reader because it seems like an impossible case.

DanAlbert commented on game/missiongenerator/aircraft/flightgroupspawner.py:
```diff
@@ -93,8 +93,7 @@ def generate_flight_at_departure(self) -> FlyingGroup[Any]:
             elif isinstance(cp, NavalControlPoint):
-                group_name = cp.get_carrier_group_name()
-                carrier_group = self.mission.find_group(group_name)
+                carrier_group = cp.get_carrier_group()
```
I don't think this is right. The `Group` in the mission and the `Group` in the TGO are not the same object. The `Group` in `TheaterGroundObject` is just a template for what to spawn in the mission. The thing in `Mission` is the actual group that will spawn. They probably have different IDs. Did you test that aircraft on carriers actually get assigned to the carrier in the generated mission? If so I don't understand why this works :(

MetalStormGhost replied:
You are right, thanks for spotting this! Fixed in 2352fbfcfef7ea43625b76fae0913733e326ed2a

DanAlbert commented on game/theater/controlpoint.py:
```diff
@@ -525,14 +532,60 @@ def get_carrier_group_name(self) -> Optional[str]:
+                                CVN_71,
+                                CVN_72,
+                                CVN_73,
+                                CVN_75,
```
AFAICT these cases don't happen. Where are these coming from?

MetalStormGhost replied:
In preparation for #1787. Supercarriers are pretty to look at even if you don't have the module, so I'm future-proofing in case someone (such as myself) wants to use both supercarrier and non-supercarrier types in the same campaign.

DanAlbert commented on game/theater/controlpoint.py:
```diff
@@ -1011,8 +1064,14 @@ def runway_is_operational(self) -> bool:
                 if db.unit_type_from_name(u.type) in [
                     Forrestal,
                     Stennis,
+                    VINSON,
```
This list shouldn't be copy pasted everywhere. Factor out `game.db.CARRIERS` and `game.db.LHAS`. It seems like you're also repeating what is effectively `NavalControlPoint.get_carrier_unit` a lot.

MetalStormGhost replied:
Done in 2352fbfcfef7ea43625b76fae0913733e326ed2a, thanks for the suggestion! Cleaned up the code quite a bit.

DanAlbert commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -241,10 +242,26 @@ def closeEvent(self, close_event: QCloseEvent):
+            ):
+                supercarrier_type = db.upgrade_to_supercarrier(
+                    db.ship_type_from_name(carrier_type), self.cp.name
+                ).id
```
Any reason to not handle this within `get_carrier_group_type`? We don't want the upgrade code copy/pasted everywhere, so probably better to do that?

MetalStormGhost replied:
I suppose not. I'll make the change.

MetalStormGhost replied:
Upon further investigation, ControlPoint doesn't have access to game. Seems like a big refactor to add that, since the control points are generated when Campaign is created and Game doesn't exist at that point yet.

MetalStormGhost committed:
Fixed a bug in flightgroupspawner.py which did not account for the fact the Group in the mission and the Group in the TGO are not the same object. Factored out carrier types and LHA types to game.db.CARRIERS and game.db.LHAS respectively. Cleaned up NavalControlPoint.runway_is_operational().

MetalStormGhost committed:
Fixed a bug in flightgroupspawner.py which did not account for the fact the Group in the mission and the Group in the TGO are not the same object. Factored out carrier types and LHA types to game.db.CARRIERS and game.db.LHAS respectively. Cleaned up NavalControlPoint.runway_is_operational().

MetalStormGhost committed:
Added a check for returned None from cp.get_carrier_group() in flightgroupspawner.py

MetalStormGhost committed:
Added a check for returned None from cp.get_carrier_group() in flightgroupspawner.py

MetalStormGhost commented:
I removed the VINSON support from game/db.py

I did not remove the base image png in case the support will be added in a future PR.

DanAlbert reviewed (changes requested):
There are open comments with no response.

MetalStormGhost committed:
Merged ConflictTheater access to controlpoint.py

RndName commented:
Most of this code is going to be obsolete whenever #1989 gets merged as we then have a ShipUnitType which allows us to check for the actual UnitClass of a unit. In this case we don't need a check against a DB of Carrier or LHA Types.
Also this code here has some changes coming from the latest Airfield changes from DanAlbert. It looks messed up and even Github shows conflicts.

I would suggest to wait for the refactor before merging this one here as it will be easier and cleaner than ever to check the actual type of the Ship.

MetalStormGhost committed:
Moved upgrade_to_supercarrier logic from tgogenerator.py to controlpoint.py

MetalStormGhost committed:
Reverted flightgroupspawner.py to ad0d3412fb09f9b5fcf7eb5d430204115aeaee99

MetalStormGhost committed:
Corrected GenericCarrierGenerator get_carrier_type() return type to Optional[ShipType] since ControlPoint might return None if the carrier group does not exist.

MetalStormGhost committed:
Cleaned up typing.

MetalStormGhost committed:
File game/theater/theatergroundobject.py, revision 73a8ec02b2c2f8026ab943b789394652fe5fcb77

MetalStormGhost committed:
Moved get_carrier_type logic from tgogenerator.py to theatergroundobject.py

MetalStormGhost committed:
Resolving a conflict in theatergroundobject.py imports.

RndName commented on game/missiongenerator/tgogenerator.py:
```diff
@@ -374,11 +367,8 @@ def generate(self) -> None:
-        if issubclass(carrier_type, ShipType):
-            return carrier_type
-        raise RuntimeError(f"First unit of TGO {group.name} is no Ship")
+    def get_carrier_type(self) -> Optional[Type[ShipType]]:
```
Just some minor thing: you can make it a property `carrier_type` if you dont pass any parameters

MetalStormGhost replied:
Done.

RndName commented on game/theater/controlpoint.py:
```diff
@@ -46,6 +60,7 @@
 from .frontline import FrontLine
 from .missiontarget import MissionTarget
 from .theatergroundobject import (
+    BuildingGroundObject,
```
Where does this come from?

MetalStormGhost replied:
Hmm, probably intentionally included this change when resolving a conflict earlier. Removed.

RndName commented on game/missiongenerator/tgogenerator.py:
```diff
@@ -360,10 +352,11 @@ def generate(self) -> None:
 
             # Set Carrier Specific Options
-            if g_id == 0:
+            carrier_type = self.get_carrier_type()
```
as only group 0 is the carrier group there is no need to get the type for any other group here. 
Also we should raise an error if this returns None as then something else went quite wrong as we are in the carrier generator so this should never ever be None

MetalStormGhost replied:
I have now changed this to match your suggestions.

RndName commented on game/theater/theatergroundobject.py:
```diff
@@ -260,6 +261,12 @@ def group_by_name(self, name: str) -> Optional[TheaterGroup]:
+        if issubclass(carrier_type, ShipType):
+            return carrier_type
+        raise RuntimeError(f"First unit of TGO {self.name} is no Ship")
+
```
this function can be removed as it's only used by the above caller anymore

MetalStormGhost replied:
Done, thanks!

RndName commented on game/theater/controlpoint.py:
```diff
@@ -608,6 +623,51 @@ def get_carrier_group_name(self) -> Optional[str]:
+                            UnitClass.HELICOPTER_CARRIER,
+                        ]:
+                            carrier_type = g.get_carrier_type()
+                            if carrier_type is not None:
```
you can replace this call and just check the first unit. For a carrier group it is a requirement that the first unit is the carrier. Then you dont have to check for None as well :)

MetalStormGhost replied:
Done. Thanks!

RndName commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -246,10 +247,14 @@ def closeEvent(self, close_event: QCloseEvent):
-            return "./resources/ui/carrier.png"
-        elif self.cp.cptype == ControlPointType.LHA_GROUP:
-            return "./resources/ui/lha.png"
+        game = self.game_model.game
```
is this used anywhere?

MetalStormGhost replied:
Nope, leftovers from the earlier upgrade_to_supercarrier call in QBaseMenu2.py
This has now been removed.

RndName reviewed (changes requested):
Only some minor things :)

MetalStormGhost committed:
get_carrier_type calls changed according to suggestions from @RndName

- as only group 0 is the carrier group there is no need to get the type for any other group here.
Also we should raise an error if this returns None as then something else went quite wrong as we are in the carrier generator so this should never ever be None
- function can be removed as it's only used by the above caller anymore
- you can replace this call and just check the first unit. For a carrier group it is a requirement that the first unit is the carrier. Then you dont have to check for None as well :)

Also removed some leftovers from previous commits.

RndName commented on game/theater/theatergroundobject.py:
```diff
@@ -260,17 +263,11 @@ def group_by_name(self, name: str) -> Optional[TheaterGroup]:
-        for unit in self.units:
-            unit.position.heading += rotation
-            unit.position.rotate(self.position, rotation)
+    def get_carrier_type(self) -> Type[ShipType]:
```
sth went wrong during rebase ;)

MetalStormGhost replied:
Indeed! I don't know what happened. This should now be fixed.

RndName commented on game/theater/controlpoint.py:
```diff
@@ -21,13 +21,27 @@
     Set,
     TYPE_CHECKING,
     Tuple,
+    Union,
```
Used?

MetalStormGhost replied:
Nope, I have removed this.

RndName commented on qt_ui/windows/basemenu/QBaseMenu2.py:
```diff
@@ -240,10 +241,12 @@ def closeEvent(self, close_event: QCloseEvent):
+            self.cp.cptype == ControlPointType.AIRCRAFT_CARRIER_GROUP
+            or self.cp.cptype == ControlPointType.LHA_GROUP
+        ):
+            carrier_type = self.cp.get_carrier_group_type()
```
If super carrier option is not enabled this will always show the base carrier (stennis) and not the one representing the Carrier chosen by the Campaign (even if it is just a name) cp.name i think.

MetalStormGhost replied:
I have now changed this so it works like in your latter option.

RndName reviewed (changes requested):
I think you may accidently delted too much during rebase

MetalStormGhost committed:
Restored rotate() which was accidentally removed from theater/theatergroundobject.py

MetalStormGhost committed:
Implemented get_carrier_group_type_by_name() in ControlPoint, which returns the (super)carrier type based on the name of the carrier group. It does not check for the supercarrier game option, but rather assumes that it's enabled. Now used to display the base image for carriers.

MetalStormGhost committed:
Removed unused Union import.

RndName commented on game/theater/controlpoint.py:
```diff
@@ -612,6 +625,81 @@ def get_carrier_group_name(self) -> Optional[str]:
+                        and issubclass(carrier_type, ShipType)
+                    ):
+                        return self.upgrade_to_supercarrier(carrier_type, self.name)
+        return None
```
Duplicate code in a fresh PR is never good. This should be either solved with a function argument or with a function wrapped around but not two functions doing the exact same logic with just one difference: the return

RndName reviewed (changes requested):
I am fine with just showing the super carrier Image to match the Unit name :)
Remove the duplication and its ready to go

MetalStormGhost committed:
Removed duplication in get_carrier_group_type_by_name()

RndName committed:
Cleanup PR
